vsintegration/src/FSharp.Editor/FSharpColorizationService.fs

```diff
@@ -0,0 +1,143 @@
+        let mutable left = 0
+        let mutable right = this.Tokens.Length - 1
+        let mutable result = None
+        while result.IsNone && right >= left do
+            let middle = (left + right) / 2
+            let middleToken = this.Tokens.[middle]
+            if middleToken.TextSpan.End <= position then
+                left <- middle + 1
+            else if middleToken.TextSpan.Start > position then
+                right <- middle - 1
+            else
+                result <- Some(middleToken)
+        result
+
+[<ExportLanguageService(typeof<IEditorClassificationService>, FSharpCommonConstants.FSharpLanguageName)>]
+type internal FSharpColorizationService() =
+
+    static let colorizationDataCache = ConditionalWeakTable<(SourceText * TextSpan * Option<string>), SourceTextColorizationData>()
+    
+    static let scanSourceText(sourceText: SourceText, textSpan: TextSpan, fileName: Option<string>, defines: string list, cancellationToken: CancellationToken): SourceTextColorizationData =
+        let mutable runningLexState = ref(0L)
+        let result = new List<ClassifiedSpan>()
+        let sourceTokenizer = FSharpSourceTokenizer(defines, fileName)
+        
+        let compilerTokenToRoslynToken(colorKind: FSharpTokenColorKind) = 
+            match colorKind with
+            | FSharpTokenColorKind.Comment -> ClassificationTypeNames.Comment
+            | FSharpTokenColorKind.Identifier -> ClassificationTypeNames.Identifier
+            | FSharpTokenColorKind.Keyword -> ClassificationTypeNames.Keyword
+            | FSharpTokenColorKind.String -> ClassificationTypeNames.StringLiteral
+            | FSharpTokenColorKind.Text -> ClassificationTypeNames.Text
+            | FSharpTokenColorKind.UpperIdentifier -> ClassificationTypeNames.Identifier
+            | FSharpTokenColorKind.Number -> ClassificationTypeNames.NumericLiteral
+            | FSharpTokenColorKind.InactiveCode -> ClassificationTypeNames.ExcludedCode 
+            | FSharpTokenColorKind.PreprocessorKeyword -> ClassificationTypeNames.PreprocessorKeyword 
+            | FSharpTokenColorKind.Operator -> ClassificationTypeNames.Operator
+            | FSharpTokenColorKind.TypeName  -> ClassificationTypeNames.ClassName
+            | FSharpTokenColorKind.Default | _ -> ClassificationTypeNames.Text
+
+        let scanNextToken(lineTokenizer: FSharpLineTokenizer, colorMap: string[], lexState: Ref<int64>) =
+            let tokenInfoOption, currentLexState = lineTokenizer.ScanToken(lexState.Value)
+            lexState.Value <- currentLexState
+            match tokenInfoOption with
+            | None -> false
+            | Some(tokenInfo) ->
+                let classificationType = compilerTokenToRoslynToken(tokenInfo.ColorClass)
+                for i = tokenInfo.LeftColumn to tokenInfo.RightColumn do
+                    Array.set colorMap i classificationType
+                true
+
+        let scanSourceLine(textLine: TextLine, lexState: Ref<int64>) =
```

Comment:
This function feels less like F# and much more like converted C# due to the parameters as a tuple, the mutable and the nested while loops. I would have expected a different approach. Or is performance the concern here partially?